src/huggingface_hub/utils/cache_explorer.py

```diff
@@ -0,0 +1,246 @@
+    r"""Information about a cached repository (dataset or model).
+
+    /!\ `size_on_disk` is not necesarily the sum of all revisions sizes because of
+        duplicated files. Also only blobs are taken into account, not the (neglectable)
+        size of folders and symlinks.
+    """
+
+    repo_id: str
+    repo_type: Literal["model", "dataset"]
+    repo_path: Path
+    revisions: Set[CachedRevisionInfo]
+    size_on_disk: int
+    nb_files: int
+
+    @property
+    def size_on_disk_str(self) -> str:
+        """Human-readable sizes"""
+        return _format_size(self.size_on_disk)
+
+
+@dataclass
+class HFCacheInfo:
+    r"""Information about the entire cache repository.
+
+    /!\ Here `size_on_disk` is equal to the sum of all repo sizes (only blobs).
+    """
+
+    datasets: List[CachedRepoInfo]
+    models: List[CachedRepoInfo]
+    size_on_disk: int
+
+    @property
+    def size_on_disk_str(self) -> str:
+        return _format_size(self.size_on_disk)
+
+
+class CorruptedCacheException(Exception):
+    """Exception for any unexpected structure in the huggingface cache."""
+
+
+def scan_cache_dir(
+    cache_dir: Optional[Union["str", Path]] = None
+) -> List[CachedRepoInfo]:
+    """Scan the entire cache directory and return information about it."""
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+
+    datasets: List[CachedRepoInfo] = []
+    models: List[CachedRepoInfo] = []
+    for repo_path in Path(cache_dir).iterdir():
+        try:
+            if repo_path.is_dir():
+                repo_info = _scan_cached_repo(repo_path)
+                if repo_info.repo_type == "dataset":
+                    datasets.append(repo_info)
+                elif repo_info.repo_type == "model":
+                    models.append(repo_info)
+                else:
+                    raise ValueError()
+        except CorruptedCacheException as e:
+            print(f"Error while scanning {repo_path}: {e}")
+
+    return HFCacheInfo(
+        datasets=sorted(datasets, key=lambda repo: repo.repo_path),
+        models=sorted(models, key=lambda repo: repo.repo_path),
+        size_on_disk=sum(repo.size_on_disk for repo in datasets + models),
+    )
+
+
+def _scan_cached_repo(repo_path: Path) -> Optional[CachedRepoInfo]:
+    """Scan a single cache repo and return information about it.
+
+    Any unexpected behavior will raise a `CorruptedCacheException`.
+    """
+    if not repo_path.is_dir():
+        raise CorruptedCacheException("Not a directory !")
+
+    if "--" not in repo_path.name:
+        raise CorruptedCacheException("Not a valid HuggingFace cache directory !")
+
+    repo_type, repo_id = repo_path.name.split("--", maxsplit=1)
+    if repo_type not in {"datasets", "models"}:
+        raise CorruptedCacheException("Repo type must be `dataset` or `model` !")
+    repo_type = repo_type[:-1]  # "models" -> "model"
+
+    blob_sizes: Dict[Path, int] = {}  # Key is blob_path, value is blob size (in bytes)
```

Comment:
Same here, the comment gives all the info needed, making the type hint irrelevant.